### PR TITLE
Vector search doc

### PR DESCRIPTION
**Pull request by sivukhin**

### CODE REVIEW

File: features/ai-and-embeddings.mdx [hunk 1]
```diff
@@ -1,22 +1,70 @@
...
+description: Vector Similarity Search is built into Turso and libSQL Server as a native feature.
 ---
 
-Turso enables vector search without an extension, it's just another datatype in SQLite.
+Turso and libSQL enables vector search capability without an extension.
 
 <Snippet file="technical-preview-banner.mdx" />
 
-## How to use
+## Hot it works
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:10]:
```suggestion
## How it works
```

Comment (notrab) [inline features/ai-and-embeddings.mdx:10]:
```suggestion
## How it works
```

File: features/ai-and-embeddings.mdx [hunk 2]
```diff
@@ -1,22 +1,70 @@
...
 
-- Use `F32_BLOB` / `FLOAT32` type names to define vector table column
-- Use `vector` function to convert vector from string representation into the binary format
-- Use `vector_extract` function to extract string representation from vector in binary format
-- Use `vector_distance_cos` function to calculate cosine distance between two vectors
-- Use `libsql_vector_idx(emb)` expression in the `CREATE INDEX` statement to create ANN-index
-- Use `vector_top_k(idx_name, q_vector, k)` [table-valued function](https://www.sqlite.org/vtab.html#table_valued_functions) to query ANN-index
+- You create a table with one or more vector columns (e.g. `FLOAT32`)
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:12]:
```suggestion
- Create a table with one or more vector columns (e.g. `FLOAT32`)
```

File: features/ai-and-embeddings.mdx [hunk 3]
```diff
@@ -1,22 +1,70 @@
...
+- You provide vector values in binary format or convert text representation to binary with the appropriate conversion function (e.g. `vector32(...)`)
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:13]:
```suggestion
- Provide vector values in binary format or convert text representation to binary using the appropriate conversion function (e.g. `vector32(...)`)
```

File: features/ai-and-embeddings.mdx [hunk 4]
```diff
@@ -1,22 +1,70 @@
...
+- You can calculate vectors similarity between vectors in the table or from the query itself with dedicated vector functions (e.g. `vector_distance_cos`)
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:14]:
```suggestion
- Calculate vector similarity between vectors in the table or from the query itself using dedicated vector functions (e.g. `vector_distance_cos`)
```

File: features/ai-and-embeddings.mdx [hunk 5]
```diff
@@ -1,22 +1,70 @@
...
+- You can create special vector index in order to speed up nearest neighbors queries (use `libsql_vector_idx(column)` expression in the `CREATE INDEX` statement to create vector index)
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:15]:
```suggestion
- Create a special vector index to speed up nearest neighbors queries (use the `libsql_vector_idx(column)` expression in the `CREATE INDEX` statement to create vector index)
```

File: features/ai-and-embeddings.mdx [hunk 6]
```diff
@@ -1,22 +1,70 @@
...
+- You can query index with special `vector_top_k(idx_name, q_vector, k)` [table-valued function](https://www.sqlite.org/vtab.html#table_valued_functions)
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:16]:
```suggestion
- Query the index with the special `vector_top_k(idx_name, q_vector, k)` [table-valued function](https://www.sqlite.org/vtab.html#table_valued_functions)
```

File: features/ai-and-embeddings.mdx [hunk 7]
```diff
@@ -1,22 +1,70 @@
...
 
-## Usage
+# Vectors
+
+### Types
+
+LibSQL uses native SQLite BLOB [storage class](https://www.sqlite.org/datatype3.html#storage_classes_and_datatypes) for vector columns. In order to align with SQLite [affinity rules](https://www.sqlite.org/datatype3.html#determination_of_column_affinity) all type names have 2 alternatives: one which is easy to type and another with `_BLOB` suffix which is consistent with affinity rules.
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:22]:
```suggestion
LibSQL uses the native SQLite BLOB [storage class](https://www.sqlite.org/datatype3.html#storage_classes_and_datatypes) for vector columns. To align with SQLite [affinity rules](https://www.sqlite.org/datatype3.html#determination_of_column_affinity), all type names have two alternatives: one that is easy to type and another with a _BLOB suffix that is consistent with affinity rules.
```

File: features/ai-and-embeddings.mdx [hunk 8]
```diff
@@ -1,22 +1,70 @@
...
+
+<Info>
+We suggest library authors to use type names with `_BLOB` suffix in order to make result more generic and universal. For the regular application, developer can choose any alternative he like as type name only plays a role of **hint** for SQLite and external extensions.
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:25]:
```suggestion
We suggest library authors use type names with the `_BLOB` suffix to make results more generic and universal. For regular applications, developers can choose either alternative, as the type name only serves as a **hint** for SQLite and external extensions.
```

File: features/ai-and-embeddings.mdx [hunk 9]
```diff
@@ -1,22 +1,70 @@
...
+</Info>
+
+<Info>
+As LibSQL do not introduce new storage class -- all metadata about vector also encoded in the `BLOB` itself. This comes at the price of few bytes per every row, but simplify design of the feature a lot.
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:29]:
```suggestion
As LibSQL does not introduce a new storage class, all metadata about vectors is also encoded in the `BLOB` itself. This comes at the cost of a few bytes per row but greatly simplifies the design of the feature.
```

File: features/ai-and-embeddings.mdx [hunk 10]
```diff
@@ -1,22 +1,70 @@
...
+</Info>
+
+The table below lists 6 vector types supported by LibSQL at the moment. Types are listed from more precise and more storage heavy to more compact but less precise alternatives (amount of dimensions in vector $D$ used to esimate storage requirements for single vector).
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:32]:
```suggestion
The table below lists six vector types currently supported by LibSQL. Types are listed from more precise and storage-heavy to more compact but less precise alternatives (the number of dimensions in vector $D$ is used to estimate storage requirements for a single vector).
```

File: features/ai-and-embeddings.mdx [hunk 11]
```diff
@@ -1,22 +1,70 @@
...
+ 
+| Type name                  | Storage (bytes)    | Description |
+| -------------------------- | ------------------ | ----------- |
+| `FLOAT64` \| `F64_BLOB`     | $8D + 1$           | implementation of [IEEE 754 double precision format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) for 64-bit floating point numbers |
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:36]:
```suggestion
| `FLOAT64` \| `F64_BLOB`     | $8D + 1$           | Implementation of [IEEE 754 double precision format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) for 64-bit floating point numbers |
```

File: features/ai-and-embeddings.mdx [hunk 12]
```diff
@@ -1,22 +1,70 @@
...
+| `FLOAT32` \| `F32_BLOB`     | $4D$               | implementation of [IEEE 754 single precision format](https://en.wikipedia.org/wiki/Single-precision_floating-point_format) for 32-bit floating point numbers |
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:37]:
```suggestion
| `FLOAT32` \| `F32_BLOB`     | $4D$               | Implementation of [IEEE 754 single precision format](https://en.wikipedia.org/wiki/Single-precision_floating-point_format) for 32-bit floating point numbers |
```

File: features/ai-and-embeddings.mdx [hunk 13]
```diff
@@ -1,22 +1,70 @@
...
+| `FLOAT16` \| `F16_BLOB`     | $2D + 1$           | implementation of [IEEE 754-2008 half precision format](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) for 16-bit floating point numbers |
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:38]:
```suggestion
| `FLOAT16` \| `F16_BLOB`     | $2D + 1$           | Implementation of [IEEE 754-2008 half precision format](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) for 16-bit floating point numbers |
```

File: features/ai-and-embeddings.mdx [hunk 14]
```diff
@@ -1,22 +1,70 @@
...
+| `FLOATB16` \| `FB16_BLOB`   | $2D + 1$           | implementation of [bfloat16 format](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) for 16-bit floating point numbers |
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:39]:
```suggestion
| `FLOATB16` \| `FB16_BLOB`   | $2D + 1$           | Implementation of [bfloat16 format](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) for 16-bit floating point numbers |
```

File: features/ai-and-embeddings.mdx [hunk 15]
```diff
@@ -1,22 +1,70 @@
...
+| `FLOAT8` \| `F8_BLOB`       | $D + 14$           | LibSQL specific implementation which compresses each vector component to single `u8` byte `b` and reconstruct value from it using simple transformation: $\texttt{shift} + \texttt{alpha} \cdot b$ |
+| `FLOAT1BIT` \| `F1BIT_BLOB` | $\lceil \frac{D}{8} \rceil + 3$ | LibSQL specific implementation which compresses each vector component down to 1-bit and pack multiple components into single machine word - thus achieving very compact representation |
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:41]:
```suggestion
| `FLOAT1BIT` \| `F1BIT_BLOB` | $\lceil \frac{D}{8} \rceil + 3$ | LibSQL-specific implementation which compresses each vector component down to 1-bit and packs multiple components into a single machine word, achieving a very compact representation |
```

File: features/ai-and-embeddings.mdx [hunk 16]
```diff
@@ -1,22 +1,70 @@
...
+
+<Info>
+For most applications `FLOAT32` type should be a good starting point, but you may want to explore more compact options if your table has huge amount of rows with vectors.
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:44]:
```suggestion
For most applications, the `FLOAT32` type should be a good starting point, but you may want to explore more compact options if your table has a large number of rows with vectors.
```

File: features/ai-and-embeddings.mdx [hunk 17]
```diff
@@ -1,22 +1,70 @@
...
+</Info>
+
+<Info>
+While `FLOAT16` and `FLOATB16` uses same amount of storage provides different tradeoffs between speed and accuracy. Genearally, operations over `bfloat16` are faster but this come in exchange to lower precision.
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:48]:
```suggestion
While `FLOAT16` and `FLOATB16` use the same amount of storage, they provide different trade-offs between speed and accuracy. Generally, operations over `bfloat16` are faster but come at the expense of lower precision.
```

File: features/ai-and-embeddings.mdx [hunk 18]
```diff
@@ -1,22 +1,70 @@
...
+</Info>
+
+### Functions
+
+In order to work with vectors LibSQL provides several functions which operates on the vector domain. Every function understand vector in binary format aligned with 6 types described above or in text format as a single JSON array of numbers.
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:53]:
```suggestion
To work with vectors, LibSQL provides several functions that operate in the vector domain. Each function understands vectors in binary format aligned with the six types described above or in text format as a single JSON array of numbers.
```

File: features/ai-and-embeddings.mdx [hunk 19]
```diff
@@ -1,22 +1,70 @@
...
+
+At the moment LibSQL supports following functions:
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:55]:
```suggestion
Currently, LibSQL supports the following functions:
```

File: features/ai-and-embeddings.mdx [hunk 20]
```diff
@@ -73,39 +124,121 @@ FROM movies;
 
 </Step>
 
-<Step title="Combine regular table data with vector search">
+</Steps>
+
+### Vector limitations
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:129]:
```suggestion
### Vector Limitations
```

File: features/ai-and-embeddings.mdx [hunk 21]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
+
+- Euclidian distance is **not supported** for 1-bit `FLOAT1BIT` vectors
+- Only vectors with no more than 65536 dimensions can be operated by LibSQL
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:132]:
```suggestion
- LibSQL can only operate on vectors with no more than 65536 dimensions
```

File: features/ai-and-embeddings.mdx [hunk 22]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
+
+## Indexing
+
+Nearest neighbors (NN) queries are popular for different type of AI-powered applications ([RAG](https://en.wikipedia.org/wiki/Retrieval-augmented_generation) uses NN queries to extract relevant information, recommendation engine can recommend items based on embeddings similarity).
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:136]:
```suggestion
Nearest neighbors (NN) queries are popular for various AI-powered applications ([RAG](https://en.wikipedia.org/wiki/Retrieval-augmented_generation) uses NN queries to extract relevant information, and recommendation engines can suggest items based on embedding similarity).
```

File: features/ai-and-embeddings.mdx [hunk 23]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
+
+LibSQL implements [DiskANN](https://turso.tech/blog/approximate-nearest-neighbor-search-with-diskann-in-libsql) algorithm in order to speed up approximate neareast neighbors queries for tables with vector colums. 
 
-You can combine vector search with regular query:
+<Note>
+DiskANN algorithm trades search accuracy for speed -- so it's expected that LibSQL query can return slightly suboptimal neighbors for tables with large amount of rows.
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:141]:
```suggestion
The DiskANN algorithm trades search accuracy for speed, so LibSQL queries may return slightly suboptimal neighbors for tables with a large number of rows.
```

File: features/ai-and-embeddings.mdx [hunk 24]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
+</Note>
+
+### Vector index
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:144]:
```suggestion
### Vector Index
```

File: features/ai-and-embeddings.mdx [hunk 25]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
+LibSQL introduces custom index type which helps to speed up neareast neighbors queries against fixed distance function (cosine similarity by default).
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:145]:
```suggestion
LibSQL introduces a custom index type that helps speed up nearest neighbors queries against a fixed distance function (cosine similarity by default).
```

File: features/ai-and-embeddings.mdx [hunk 26]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
+
+From the syntax perspective, vector index differs from ordinary application-defined B-Tree indices in that it must wrap vector column into `libsql_vector_idx` marker function like this:
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:147]:
```suggestion
From a syntax perspective, the vector index differs from ordinary application-defined B-Tree indices in that it must wrap the vector column into a `libsql_vector_idx` marker function like this
```

File: features/ai-and-embeddings.mdx [hunk 27]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
-LIMIT 3;
+CREATE INDEX movies_idx ON movies (libsql_vector_idx(embedding));
 ```
 
-This query demonstrates how to combine traditional SQL filtering (`year >= 2020`) with vector similarity search.
+<Note>
+Vector index works only for column with one of the vector types described above
+</Note>
+
+Vector index is fully integrated into LibSQL core -- so it inherits all operations and most features from ordinary indices:
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:157]:
```suggestion
The vector index is fully integrated into the LibSQL core, so it inherits all operations and most features from ordinary indices:
```

File: features/ai-and-embeddings.mdx [hunk 28]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
 
-</Step>
+- Index created for table with some data will be **automatically** populated with these data
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:159]:
```suggestion
- An index created for a table with existing data will be automatically populated with this data
```

File: features/ai-and-embeddings.mdx [hunk 29]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
+- All updates to the base table will be **automatically** reflected in the index
+- You can rebuild index from scratch using `REINDEX movies_idx` command
+- You can drop index with `DROP movies_idx` command
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:162]: Should this be `DROP INDEX`?

```suggestion
- You can drop index with `DROP INDEX movies_idx` command
```

File: features/ai-and-embeddings.mdx [hunk 30]
```diff
@@ -73,39 +124,121 @@ FROM movies;
...
+- You can create [partial](https://www.sqlite.org/partialindex.html) vector index with custom filtering rule:
```
Comment (notrab) [inline features/ai-and-embeddings.mdx:163]:
```suggestion
- You can create [partial](https://www.sqlite.org/partialindex.html) vector index with a custom filtering rule:
```

**Review by notrab — CHANGES_REQUESTED** [review]
@sivukhin looks great! I took an initial pass over the copy to bring it inline with the usual writing style.

File: features/ai-and-embeddings.mdx [hunk 31]
```diff
@@ -1,22 +1,70 @@
...
+- Calculate vector similarity between vectors in the table or from the query itself using dedicated vector functions (e.g. `vector_distance_cos`)
+- Create a special vector index to speed up nearest neighbors queries (use the `libsql_vector_idx(column)` expression in the `CREATE INDEX` statement to create vector index)
+- Query the index with the special `vector_top_k(idx_name, q_vector, k)` [table-valued function](https://www.sqlite.org/vtab.html#table_valued_functions)
 
-## Usage
+# Vectors
+
+### Types
+
+LibSQL uses the native SQLite BLOB [storage class](https://www.sqlite.org/datatype3.html#storage_classes_and_datatypes) for vector columns. To align with SQLite [affinity rules](https://www.sqlite.org/datatype3.html#determination_of_column_affinity), all type names have two alternatives: one that is easy to type and another with a _BLOB suffix that is consistent with affinity rules.
```
Comment (penberg) [inline features/ai-and-embeddings.mdx:22]: Add backticks to `_BLOB` here?

File: features/ai-and-embeddings.mdx [hunk 32]
```diff
@@ -1,22 +1,70 @@
...
+| `FLOAT16` \| `F16_BLOB`     | $2D + 1$           | Implementation of [IEEE 754-2008 half precision format](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) for 16-bit floating point numbers |
+| `FLOATB16` \| `FB16_BLOB`   | $2D + 1$           | Implementation of [bfloat16 format](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) for 16-bit floating point numbers |
+| `FLOAT8` \| `F8_BLOB`       | $D + 14$           | LibSQL specific implementation which compresses each vector component to single `u8` byte `b` and reconstruct value from it using simple transformation: $\texttt{shift} + \texttt{alpha} \cdot b$ |
+| `FLOAT1BIT` \| `F1BIT_BLOB` | $\lceil \frac{D}{8} \rceil + 3$ | LibSQL-specific implementation which compresses each vector component down to 1-bit and packs multiple components into a single machine word, achieving a very compact representation |
+
+<Info>
+For most applications, the `FLOAT32` type should be a good starting point, but you may want to explore more compact options if your table has a large number of rows with vectors.
```
Comment (penberg) [inline features/ai-and-embeddings.mdx:44]: Would `FLOAT16` be a better default to recommend? It's smaller but still has plenty of precision?

Comment (sivukhin) [reply features/ai-and-embeddings.mdx:44]: Float16 can be relatively slow + it's harder to insert float16 blobs from clients (but with float32 users can just insert blobs from Float32Array)

**Review by notrab — APPROVED** [review]
:shipit:

**Pull request merged**